### ISSUE TITLE
add @postConstruct to the lifecycle

### ISSUE THREAD

**Issue by dcavanagh** [inversify/InversifyJS#587]
I have added @postConstruct method decorator that will run after an object is instantiated like in [Spring](https://docs.spring.io/spring/docs/current/spring-framework-reference/htmlsingle/#beans-postconstruct-and-predestroy-annotations)

## Expected Behavior
if you decorate a method with @postConstruct that method will be called after the object is constructed

## Current Behavior
This behavior does not currently exist

## Possible Solution
I am working on a change and have it working but it needs some review and maybe some refactoring to make it fit in inversify better before i submit a pull request. [fork](https://github.com/AltekkeE/InversifyJS)

## Steps to Reproduce (for bugs)
N/A

## Context
Keep object setup out of the constructor

## Your Environment

* Version used: 4.1.1
* Environment name and version (e.g. Chrome 39, node.js 5.4): Node v7.4.0
* Operating System and version (desktop or mobile): OSX
* Link to your project:  [fork](https://github.com/AltekkeE/InversifyJS)

# Stack trace
N/A

**Comment by remojansen** [comment]
Hi @AltekkeE I like this feature :+1: I have some questions if you don't mind... 

How do you think this should play with the `onActivation` feature? should  `@PostConstruct`  be invoked before or after the `onActivation` handler? 

I'm thinking that maybe after because we could have the `onActivation` for all implementations of a Type and `@PostConstruct` for one of the implementations. Do you agree?

**Comment by dcavanagh** [comment]
I agree! `@postConstruct` is meant to allow for implementation specific setup. I'm not sure about the order since the `onActivation` feature can change the behavior of the class if understand it correctly.

**Comment by remojansen** [comment]
True, thanks for bringing that to my memory. I think `@postConstruct` should go first then. So the order should be:

1. `const x = new X()` // new instance
2. `x.postConstruct()` // instance initialized
3. `onActivationHandler(x)` // 

I also think that we should wrap `postConstruct` with a try catch and if an error takes place throw a custom error. Something like:

> `@postConstruct` error in class X: The details of the original exception goes here.

Do you have any other concerns?

**Comment by dcavanagh** [comment]
Sounds good.. I will add that

My only other concern is that i did not add any code to the annotation phase or the plan phase. 
The resolution phase just checks to see if the `POST_CONSTRUCT` metadata exists on the prototype and then executes the method that was decorated

**Comment by gagle** [comment]
preDestroy could be useful too

https://github.com/inversify/InversifyJS/issues/551

**Comment by remojansen** [comment]
@gagle the problem is that we know when an object is created but we cannot know when it is destroyed

**Comment by dcavanagh** [comment]
@gagle @remojanse 
I was thinking about the preDestroy and came up with 2 possibilities.

1. Only allow @preDestory to be applied to classes that are in the singleton scope. Throw an error if not. Not sure if this would really be helpful because I haven't ever needed to destroyed a singleton.

2. Maybe have the container maintain a reference to the object and have the consumer notify the container when it no longer needs it? I guess that would be coding by convention and would probably just be easier to have the consumer run the method

**Comment by gagle** [comment]
Sounds ok for me, this is what I exactly thought when I created the other issue about pre and post. IF the componet is a singleton, there is already a reference to the instance, so we can call the predestroy when the container is closed. Conceptually, if the container creates instances and they are singleton, it is also its responsibility to manage the full lifecycle and not only the instance creation. The predestroy scenario typically won't never happen because the  developers usually don't care about destroying things, sockets and i/o things are automatically closed by the underlying system and resources are freed up, but it's more cleaner if connections can be gracefully closed. In other words, the framework should allow to free up resources, it's not the framework's responsibility to close or not to close resources, so let's enable this feature to the developer.

**Comment by remojansen** [comment]
@AltekkeE I'm sorry but I won't be able to do a release today but I will get it out during the weekend. 

About preDestroy. Let's think about the API before we implement it. 

I have plans for custom scopes and the API would be something like:

In inversify-express-utils, when a request hits the server:

```ts
const requestScope = container.createScope("requestScope") // Symbol("requestScope");

const scope = container.beginScope(requestScope); // creates a new scope instance
const controllerInstance = container.getInScope("Controller", scope);
// ...
container.endScope(scope); // kills the controller instance invokes @predestory
```

We would need `getInScope`, `getInScopeNamed`, `getAllInScope` & `getInScopeTagged` ...

User will be able to do:

```ts
container.bind<Controller>("Controller").to(SomeController).inCustomScope(requestScope);
container.bind<Repository>("Repository").to(SomeRepository).inCustomScope(requestScope);
```

Internally the container will keep dictionary like:

```ts
scopeID // Symbol("requestScope")
scopeInstanceID // Some auto-generated ID
valuesArray // An array that contains all the objects in the scope isntance
```

**Comment by sanex3339** [comment]
I like this.

In my project i need to pass all required dynamic values through `initialize` method, for example:
```typescript
class Foo implements IFoo {
    @initializable()
    private valueA;

    @initializable()
    private valueB;

    constructor(@inject(...) bar: IBar) {...
    }

    public initialize(valueA: number, valueB: number) {
        this.valueA = valueA;
        this.valueB = valueB;

        // will be the runtime Error if valueA or valueB will be undefined at the end of `initialize` method
    }

    getSum() {
        return this.valueA + this.valueB;
    }
}

class App {
    private foo: IFoo;

    constructor(@inject(...) foo: IFoo) {
        this.foo = foo;
    }

    getMathRandomSum() {
        this.foo.initialize(Math.random(), Math.random());

        return this.foo.getSum();
    }
}
```

Decorator:
https://github.com/javascript-obfuscator/javascript-obfuscator/blob/master/src/decorators/Initializable.ts

Method call:
https://github.com/javascript-obfuscator/javascript-obfuscator/blob/master/src/custom-nodes/string-array-nodes/group/StringArrayCustomNodeGroup.ts#L122

Method:
https://github.com/javascript-obfuscator/javascript-obfuscator/blob/master/src/custom-nodes/string-array-nodes/StringArrayNode.ts#L53

So, i really want some way to make this things without custom decorators, only with Inversify.js decorators.

**Comment by dcavanagh** [comment]
@remojansen no problem! thanks for all your help

**Comment by remojansen** [comment]
Hi guys, I'm closing this issue I have created https://github.com/inversify/InversifyJS/issues/593 to follow @preDestroy  I just released 4.2.0
